### PR TITLE
Avogadro flask

### PR DESCRIPTION
**Pull request by bnmajor**
Move Avogadro functions into a flask container

### CODE REVIEW

File: flask/avogadro/src/server.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
+import json
+
+from flask import Flask, request
+
+import avogadro_api as avogadro
+
+app = Flask(__name__)
+
+
+@app.route('/calculate', methods=['POST'])
```
Comment (cjh1) [inline flask/avogadro/src/server.py]: We probably want to be a little more explicit with the name `/calculate-mo`?

File: flask/avogadro/src/server.py [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+def calculate():
+    json_data = request.get_json()
+    cjson = json_data['cjson']
```
Comment (cjh1) [inline flask/avogadro/src/server.py]: Lets add a little validation here. Checking the JSON contains the 'cjson' and 'mo' keys and return a status 400 if it doesn't.

File: flask/avogadro/src/server.py [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    mo = json_data['mo']
+
+    return avogadro.calculate_mo(cjson, mo)
+
+
+@app.route('/convert-str/<output>', methods=['POST'])
+def convert_string(output):
+    json_data = request.get_json()
+    input_format = json_data['format']
```
Comment (cjh1) [inline flask/avogadro/src/server.py]: Again a little validation.

File: flask/avogadro/src/server.py [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+    data = json.dumps(json_data['data'])
+
+    return avogadro.convert_str(data, input_format, output)
+
+
+@app.route('/properties/<property_type>', methods=['POST'])
+def get_properties(property_type):
+    json_data = request.get_json()
+    input_format = json_data['format']
```
Comment (cjh1) [inline flask/avogadro/src/server.py]: Here as well.

File: girder/molecules/molecules/avogadro.py [hunk 5]
```diff
@@ -1,34 +1,61 @@
...
+    base_url = avogadro_base_url()
+    path = 'properties'
+    url = '/'.join([base_url, path, 'atom'])
+
+    data = {
+        'format': in_format,
+        'data': str_data,
+    }
+
+    r = requests.post(url, json=data)
```
Comment (cjh1) [inline girder/molecules/molecules/avogadro.py]: We should add a [raise_for_status](https://2.python-requests.org/en/master/api/#requests.Response.raise_for_status), so we can see when the requests fails.

File: girder/molecules/molecules/avogadro.py [hunk 6]
```diff
@@ -1,34 +1,61 @@
...
+    base_url = avogadro_base_url()
+    path = 'properties'
+    url = '/'.join([base_url, path, 'molecule'])
+
+    data = {
+        'format': in_format,
+        'data': str_data,
+    }
+
+    r = requests.post(url, json=data)
```
Comment (cjh1) [inline girder/molecules/molecules/avogadro.py]: raise_for_status

File: girder/molecules/molecules/avogadro.py [hunk 7]
```diff
@@ -101,24 +128,17 @@ def calculation_properties(json_data):
...
+    base_url = avogadro_base_url()
+    path = 'calculate'
+    url = '/'.join([base_url, path])
+
+    data = {
+        'cjson': cjson,
+        'mo': mo,
+    }
+
+    r = requests.post(url, json=data)
```
Comment (cjh1) [inline girder/molecules/molecules/avogadro.py]: raise_for_status

File: girder/molecules/molecules/utilities/async_requests.py [hunk 8]
```diff
@@ -120,3 +124,39 @@ def _finish_3d_coords_gen(inchikey, user, future):
...
+    data = {
+        'cjson': cjson,
+        'mo': mo,
+    }
+
+    session = FuturesSession()
+    future = session.post(url, json=data)
+
+    future.add_done_callback(functools.partial(
+        _finish_orbital_gen, mo, id, getCurrentUser(), orig_mo))
```
Comment (cjh1) [inline girder/molecules/molecules/utilities/async_requests.py]: Probably want some error handling here as well, not sure if  requests-future has a `raise_for_status`?

**Comment by cjh1** [comment]
@bnmajor I have build and pushed to the container. I also updated the CircleCI configuration to start the avogadro container to fix the tests.

**Comment by cjh1** [comment]
Still broken, need to check how CircleCI managed docker network interfaces, seems like OpenBabel and Avogadro are fighting over the same address.

File: flask/avogadro/src/server.py (hunk 1 above)
Comment (bnmajor) [reply flask/avogadro/src/server.py]: Fixed.

**Comment by bnmajor** [comment]
Added error handling for synchronous requests, asynchronous is still not handled. The test for getting the calculation cube has been added to `plugin_tests`.

**Comment by cjh1** [comment]
@bnmajor You new test seems to be failing on CircleCI can you take a look?

**Pull request merged**